lib/concurrent/executor/ruby_thread_pool_executor.rb

```diff
@@ -123,142 +122,231 @@ def queue_length
+      mutex.synchronize do
+        if ns_limited_queue?
+          @max_queue - @queue.length
+        else
+          -1
+        end
+      end
     end
 
-    # Returns an array with the status of each thread in the pool
-    #
-    # This method is deprecated and will be removed soon.
-    def status
-      warn '[DEPRECATED] `status` is deprecated and will be removed soon.'
-      mutex.synchronize { @pool.collect { |worker| worker.status } }
+    alias_method :current_length, :length # TODO remove?
+
+    # @api private
+    def remove_busy_worker(worker)
+      mutex.synchronize { ns_remove_busy_worker worker }
     end
 
-    # Run on task completion.
-    #
-    # @!visibility private
-    def on_end_task
-      mutex.synchronize do
-        @completed_task_count += 1 #if success
-        break unless running?
-      end
+    # @api private
+    def ready_worker(worker)
+      mutex.synchronize { ns_ready_worker worker }
     end
 
-    # Run when a thread worker exits.
-    #
-    # @!visibility private
-    def on_worker_exit(worker)
-      mutex.synchronize do
-        @pool.delete(worker)
-        if @pool.empty? && !running?
-          stop_event.set
-          stopped_event.set
-        end
-      end
+    # @api private
+    def worker_not_old_enough(worker)
+      mutex.synchronize { ns_worker_not_old_enough worker }
+    end
+
+    # @api private
+    def worker_died(worker)
+      mutex.synchronize { ns_worker_died worker }
     end
 
     protected
 
-    # @!visibility private
-    def execute(*args, &task)
-      if ensure_capacity?
+    def ns_limited_queue?
+      @max_queue != 0
+    end
+
+    def ns_execute(*args, &task)
+      if ns_assign_worker(*args, &task) || ns_enqueue(*args, &task)
         @scheduled_task_count += 1
-        @queue << [args, task]
       else
-        handle_fallback(*args, &task) if @max_queue != 0 && @queue.length >= @max_queue
+        handle_fallback(*args, &task)
       end
-      prune_pool
+
+      ns_prune_pool if @next_gc_time < Concurrent.monotonic_time
+      # raise unless @ready.empty? || @queue.empty? # assert
     end
 
-    # @!visibility private
-    def shutdown_execution
+    alias_method :execute, :ns_execute
+
+    def ns_shutdown_execution
       if @pool.empty?
+        # nothing to do
         stopped_event.set
-      else
-        @pool.length.times { @queue << :stop }
       end
+      if @queue.empty?
+        # no more tasks will be accepted, just stop all workers
+        @pool.each(&:stop)
+      end
+
+      # raise unless @ready.empty? || @queue.empty? # assert
     end
 
-    # @!visibility private
-    def kill_execution
-      @queue.clear
-      drain_pool
+    alias_method :shutdown_execution, :ns_shutdown_execution
+
+    # @api private
+    def ns_kill_execution
+      ns_shutdown_execution
+      unless stopped_event.wait(1)
+        @pool.each &:kill
+        @pool.clear
+        @ready.clear
+        # TODO log out unprocessed tasks in queue
```

Comment:
Only a debug level logging would make sense probably, when user has something not executing it may be because of premature termination so I would like to inform him.